src/snowflake/connector/query_context_cache.py

```diff
@@ -0,0 +1,302 @@
+        return self._capacity
+
+    def _add_qce(self, qce: QueryContextElement) -> None:
+        self._tree_set.add(qce)
+        self._id_map[qce.id] = copy.deepcopy(qce)
+        self._intermediate_priority_map[qce.priority] = copy.deepcopy(qce)
+
+    def _remove_qce(self, qce: QueryContextElement) -> None:
+        self._id_map.pop(qce.id)
+        self._priority_map.pop(qce.priority)
+        self._tree_set.remove(qce)
+
+    def _replace_qce(
+        self, old_qce: QueryContextElement, new_qce: QueryContextElement
+    ) -> None:
+        self._remove_qce(old_qce)
+        self._add_qce(new_qce)
+
+    def _sync_priority_map(self):
+        """
+        Sync the _intermediate_priority_map with the _priority_map at the end of the current round of merges.
+        """
+        logger.debug(
+            f"syncPriorityMap called priority_map size = {len(self._priority_map)}, new_priority_map size = {len(self._intermediate_priority_map)}"
+        )
+
+        self._priority_map.update(self._intermediate_priority_map)
+        # Clear the _intermediate_priority_map for the next round of QCC merge (a round consists of multiple entries)
+        self._intermediate_priority_map.clear()
+
+    def merge(self, id: int, read_timestamp: int, priority: int, context: str) -> None:
+        if id in self._id_map:
+            qce = self._id_map[id]
+            if (read_timestamp > qce.read_timestamp) or (
+                read_timestamp == qce.read_timestamp and priority != qce.priority
+            ):
+                # when id if found in cache and we are operating on a more recent timestamp. We do not update in-place here.
+                new_qce = QueryContextElement(id, read_timestamp, priority, context)
+                self._replace_qce(qce, new_qce)
+        else:
+            new_qce = QueryContextElement(id, read_timestamp, priority, context)
+            if priority in self._priority_map:
+                old_qce = self._priority_map[priority]
+                self._replace_qce(old_qce, new_qce)
+            else:
+                self._add_qce(new_qce)
+
+    def check_cache_capacity(self) -> None:
+        logger.debug(
+            f"check_cache_capacity() called. treeSet size is {len(self._tree_set)} and cache capacity is {self.capacity}"
+        )
+
+        while len(self._tree_set) > self.capacity:
+            # remove the qce with highest priority value => element with least priority
+            qce = self._tree_set[-1]
+            self._remove_qce(qce)
+
+        logger.debug(
+            f"check_cache_capacity() returns. treeSet size is {len(self._tree_set)} and cache capacity is {self.capacity}"
+        )
+
+    def clear_cache(self) -> None:
+        logger.debug("clear_cache() called")
+        self._id_map.clear()
+        self._priority_map.clear()
+        self._tree_set.clear()
+
+    def _get_elements(self) -> set[QueryContextElement]:
+        return self._tree_set
+
+    def _last(self) -> QueryContextElement:
```

Comment:
We are using this and `_get_elements` in the tests. I would like to keep it in case we change the underlying implementation. This way we would still keep these functions as they help test against interfaces instead of implementations.